### PR TITLE
[Bug]: Fix optimizer.zero_grad() order in compression distillation — incorrect gradient accumulation

### LINKED ISSUE #483: [Bug]: optimizer.zero_grad() called after optimizer.step() in compression distillation — incorrect gradient accumulation
## Problem / Motivation

In 
ightmarenet/compression/distillation.py:162-164, the training step order is:

`python
loss.backward()     # computes gradients (ADDING to existing .grad)
optimizer.step()    # applies gradients (which include stale gradients from previous step!)
optimizer.zero_grad()  # zeroes gradients TOO LATE
`

Standard PyTorch practice is zero_grad() → ackward() → step(). Because zero_grad() is called *after* step(), gradients from iteration N are **never cleared** before iteration N+1's ackward(). This means ackward() on iteration N+1 **accumulates** into the un-zeroed gradients from iteration N, and step() on iteration N+1 uses the sum of N and N+1 gradients.

This effectively implements **unintentional gradient accumulation** where the accumulation window grows unbounded (the window for step N includes all prior steps' gradients since the last zero_grad() — which never happens between steps).

### Impact
Model training in the compression/distillation phase has incorrect gradient values, leading to suboptimal or diverging weight updates. This affects every pipeline run that uses compression.

### Affected Files
- 
ightmarenet/compression/distillation.py:162-164

## Proposed Solution
Swap lines 162 and 164 so the order is:
`python
optimizer.zero_grad()  # NEW: clear gradients from previous step
loss.backward()        # compute fresh gradients
optimizer.step()       # apply fresh gradients
`

## Acceptance Criteria
- [ ] zero_grad() executes before ackward() in the distillation training step
- [ ] Training loss converges as expected (compare with baseline)
- [ ] All existing compression/distillation tests pass

## Scope / Difficulty Estimate
Easy — 15 min. Single-line swap. However, retraining validation is needed to confirm loss behavior changes as expected.

### PR DESCRIPTION
**Pull request by Prathvikmehra**
## Description

Fixes incorrect gradient accumulation order in the compression distillation training step.

### Problem

In `distillation.py:162–164`, the call order was:

```
loss.backward()      # computes gradients (ADDING to stale .grad)
optimizer.step()     # applies stale+new gradients together
optimizer.zero_grad() # clears too late — next iteration accumulates
```

This meant:
1. `grad` from iteration N was never cleared before iteration N+1
2. `backward()` on iteration N+1 accumulated into stale gradients
3. `step()` on iteration N+1 applied the sum of N and N+1 gradients
4. The accumulation window grew unbounded — effectively **unintentional gradient accumulation** with an ever-growing window.

Additionally, the AMP (gradient scaler) branch had **no `zero_grad()` call at all**, meaning it suffered from the same unbounded accumulation.

### Fix

Moved `optimizer.zero_grad()` to the top of each branch (before `backward()`), so the correct order is:

**Non-scaler path:**
```
optimizer.zero_grad()  # clear stale gradients
loss.backward()        # compute fresh gradients
optimizer.step()       # apply fresh gradients
```

**AMP/scaler path:**
```
optimizer.zero_grad()              # clear stale gradients
scaler.scale(loss).backward()      # compute scaled gradients
scaler.unscale_(optimizer)
scaler.step(optimizer)
scaler.update()
```

### Changes

| File | Change |
|------|--------|
| `nightmarenet/compression/distillation.py:156–164` | Move `zero_grad()` before `backward()` in both scaler and non-scaler paths |

### Verification

- [x] `zero_grad()` executes before `backward()` in both code paths
- [x] Scaler branch now has explicit gradient clearing (was missing entirely)
- [x] Ruff lint passes
- [x] Existing distillation tests pass

Closes #483

### CODE REVIEW

**Comment by Adit-Jain-srm** [comment]
**Closing. This is not a bug.**

The existing pattern `loss.backward() -> optimizer.step() -> optimizer.zero_grad()` is standard PyTorch. `zero_grad()` after `step()` clears gradients for the NEXT iteration — it does NOT cause gradient accumulation across steps.

Both "zero before backward" and "zero after step" are equivalent for single-loss-per-step training. The current code is correct. The issue this PR claims to fix does not exist.

Reference: [PyTorch docs](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html) — `zero_grad()` placement is a style choice when not accumulating gradients intentionally.

Star/follow also not completed.
